### DISCUSSION TITLE
Cannot get environment variables to work

### DISCUSSION THREAD

**Discussion by mikevang** [brefphp/bref#1194] [Support]
Please help, I'm struggling with environment variables here. My implementation calls an external API and I need to store the url and access key in an environment somewhere. I've tried using github secrets and a CD action that's supposed to put the secret into the environment, but when I run it, I get `null` for the value. After many different attempts I abandoned that and switched to AWS SSM parameter store, per the documentation, but I'm still getting `null` back from trying to access those keys.

Here's my current setup:

in AWS parameter store - `/XYYApiService-dev/API_URL`

in serverless.yml
```
provider:
  ...
  environment:
    XYZ_API_URL: ${ssm:/XYYApiService-dev/API_URL}
    ...
```

in my /config/app.php
```
'XYZ_API_URL' => env('XYZ_API_URL'),
```

and finally in the constructor for my API connector class
```
        $baseUrl = config('app.XYZ_API_URL');
        if (empty($baseUrl)) {
            error_log("Unable to retrieve url from configuration file, trying env.");
            $baseUrl = env('XYZ_API_URL');
            if (empty($baseUrl)) {
                error_log("Unable to retrieve url from local env, exiting.");
                throw new Exception("Unable to retrieve url from configuration files. [$baseUrl]");
            }
        }
```

When run, I see both `error_log` messages in the Lambda log, and it throws the exception every time. What am I missing here?

Thanks.

**Comment by mnapoli** [comment]
I would check the generated CloudFormation in the `.serverless` directory. You can also check the Lambda function configuration in the AWS console to see what value is on the environment variable.

**Reply by mikevang** [reply]
> I would check the generated CloudFormation in the `.serverless` directory. You can also check the Lambda function configuration in the AWS console to see what value is on the environment variable.

The variables show up correctly in both the cloudformation file and the lambda function config.

`cloudformation-template-update-stack.json`
```
"WebLambdaFunction": {
      ...
        "Environment": {
          "Variables": {
            ...
            "XYZ_API_URL": "<the url>",
            ...
          }
        },
```

![AWS_Lambda_config](https://user-images.githubusercontent.com/33760915/164044875-98470d64-2e3e-4b05-98d3-3096fd1e2bd5.png)

**Comment by t-richard** [comment]
Then it should be populated correctly and accessible from your lambda function code.

Are you able to test with both `getenv('XYZ_API_URL')` and `$_SERVER['XYZ_API_URL']`. Do they contain the value you are looking for?

**Reply by mikevang** [reply]
> Then it should be populated correctly and accessible from your lambda function code.
> 
> Are you able to test with both `getenv('XYZ_API_URL')` and `$_SERVER['XYZ_API_URL']`. Do they contain the value you are looking for?

I added checks for both of those, and they both fail as well. The ONLY way I've been able to get it to work so far is to add it to my local .env file and deploy with `serverless deploy`. However that will not work with the CI/CD deployment action we've setup with github since the .env file is not part of the repo.

**Comment by mikevang** [comment]
Well... that's just odd. I haven't changed my code any, aside from adding those two new checks above, which didn't work initially, but now it all of a sudden started working with the `config('app.XYZ_API_URL')` call. Is it possible that the AWS permission store takes a while to update or something along those lines?

**Reply by deleugpn** [reply]
It may be that you're uploading an `.env` file into AWS Lambda and it is taking effect and/or you might have had `php artisan config:cache` on your local machine which got uploaded into AWS.
